### PR TITLE
[SPARK-47358][SQL][COLLATION] Improve repeat expression support to return correct datatype

### PR DESCRIPTION
**Pull request by miland-db**
### What changes were proposed in this pull request?
Enable collation support for the StringRepeat built-in string function in Spark.

### Why are the changes needed?
So we can propagate input string datatype as a result of repeat expression.

### Does this PR introduce _any_ user-facing change?
Yes, users will now get different type when using REPEAT on collated strings in Spark SQL queries.

### How was this patch tested?
Unit test in `CollationStringExpressionsSuite.scala`

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 1]
```diff
@@ -70,6 +70,34 @@ class CollationStringExpressionsSuite extends QueryTest with SharedSparkSession
     })
   }
 
+  case class RepeatCollationTestCase[R](s: String, n: Int, collation: String, expectedResult: R)
+
+  test("Support Repeat string expression with Collation") {
+    // Supported collations
+    val checksRepeat = Seq(
+      RepeatCollationTestCase("Spark", 2, "UTF8_BINARY", "SparkSpark"),
```
Comment (stefankandic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:78]: i see that you're just doing the same thing as the previous test but I'm not a fan of creating a class for each test

something i would recommend would be:
```
  checksRepeat.foreach { case (s, n, collation, expectedResult) =>
    ...
  }
```

but this is not a big deal, so it's up to you whichever you prefer

**Review by stefankandic — APPROVED** [review]
nit: maybe try formatting your changes (shortcut is option+cmd+l on macos - this will format the entire file, but if you use shift as well you can choose to reformat only uncommitted changes)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -71,8 +71,7 @@ import org.apache.spark.util.ArrayImplicits._
   since = "1.5.0",
   group = "string_funcs")
 // scalastyle:on line.size.limit
-case class ConcatWs(children: Seq[Expression])
-  extends Expression with ImplicitCastInputTypes {
+case class ConcatWs(children: Seq[Expression]) extends Expression with ImplicitCastInputTypes {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:74]: Can we remove those unrelated changes? those changes make it harder to backport, cherry-pick, revert, etc.

Comment (miland-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:74]: I can do that, but all scala style tests are failing then.

**Comment by miland-db** [comment]
@MaxGekk @cloud-fan could you please check this and merge it if everything is ok?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 3]
```diff
@@ -70,6 +71,34 @@ class CollationStringExpressionsSuite extends QueryTest with SharedSparkSession
...
+
+    // Check return type
+    val checksReturnType = Seq(
+      RepeatCollationTestCase("Spark", 2, "UTF8_BINARY", "UTF8_BINARY"),
+      RepeatCollationTestCase("Spark", 2, "UTF8_BINARY_LCASE", "UTF8_BINARY_LCASE"),
+      RepeatCollationTestCase("Spark", 2, "UNICODE", "UNICODE"),
+      RepeatCollationTestCase("Spark", 2, "UNICODE_CI", "UNICODE_CI")
+    )
+    checksReturnType.foreach(ct => {
+      checkAnswer(sql(s"SELECT collation(repeat(collate('${ct.s}', '${ct.collation}'), ${ct.n}))"),
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:97]: Similar to @cloud-fan comment at https://github.com/apache/spark/pull/45611#discussion_r1533994036. Let's write unit tests, and don't test entire stack by running end-to-end tests.

Comment (miland-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala]: I have changed the tests now. Please check it again

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 4]
```diff
@@ -70,6 +73,19 @@ class CollationStringExpressionsSuite extends QueryTest with SharedSparkSession
     })
   }
 
+  test("REPEAT check output type on explicitly collated string") {
+    def testRepeat(expected: String, stringType: Integer, input: String, n: Integer): Unit = {
+      val s = Literal.create(input, StringType(stringType))
+
+      checkEvaluation(Collation(StringRepeat(s, Literal.create(n))).replacement, expected)
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:80]: Does `checkEvaluation` check the output type?

Comment (miland-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:80]: Yes, I test if the output type matches the expected result. You can see below that I test if the result has type for example `expected="UNICODE_CI`

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 5]
```diff
@@ -70,6 +73,19 @@ class CollationStringExpressionsSuite extends QueryTest with SharedSparkSession
     })
   }
 
+  test("REPEAT check output type on explicitly collated string") {
+    def testRepeat(expected: String, stringType: Integer, input: String, n: Integer): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:77]:
```suggestion
    def testRepeat(expected: String, collationId: Integer, input: String, n: Integer): Unit = {
```

**Review by cloud-fan — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Review by cloud-fan — APPROVED** [review]
